### PR TITLE
Add 'branch' parameter for a put step

### PR DESCRIPTION
**Pull request by hy0044**
This PR adds `branch` parameter for a `put` step.
It is useful to override `branch` source configuration dynamically and to push commits to different branch from one configured in source configuration.

### CODE REVIEW

**Comment by aoldershaw** [comment]
Out of curiosity, what's your use case for this feature? If the branch you wanted to push to was static, you could just have multiple resources (one for each branch). From the way it's implemented, though, it seems like you're looking for a dynamic way to decide on the branch.

This seems related to #327 - have you seen that PR?

**Comment by hy0044** [comment]
@aoldershaw 
Thank you for your comments.
My use case for this feature is release process like below. 

```
1. Create working branch for release (e.g. releng-1.0.0)
2. Push commits of pre-process to releng-1.0.0
3. Create release tag (e.g. 1.0.0)
4. Start release task on Concourse CI
  4.1 git-resource detects 1.0.0 tag 
  4.2 Create release packages
  4.3 Create release page
  4.4 Push commits of post-process to releng-1.0.0 <- this feature can be used for this step.
5. Merge commits on releng-1.0.0 back to master
```
I checked #327, but it doesn't seem to support pushing commits to a specified branch.

**Comment by aoldershaw** [comment]
I haven't looked at #327 in too much depth, but I interpreted the `checkout` param as "some git ref that you want to push to" - however, that would imply the ref would already have to exist (i.e. you can't create a new branch) - correct me if I'm wrong @xtremerui

Thanks for sharing your use case. We follow a similar flow for releasing Concourse - although, we have one pipeline per release (e.g. https://ci.concourse-ci.org/teams/main/pipelines/release-6.5.x) which has a git resource that points to the release branch.

However, we currently create our release branches manually - I guess that's what this PR is attempting to solve. The one thing that makes me a bit concerned is that the version created by the `put` would not be detected by a subsequent `check`, as the `check` is still monitoring the `source` branch. I don't totally know the implications of this - maybe @vito can shed some light on whether this is dangerous.

Another solution would be to have a `git-branch` resource type, where `put` creates a new branch. This is something that will be possible with [prototypes](https://github.com/concourse/rfcs/blob/master/037-prototypes/proposal.md) - the git prototype could support this type of operation easily - but that's down the line.

**Comment by xtremerui** [comment]
[This answer](https://stackoverflow.com/questions/38964299/why-does-concourse-get-a-resource-after-puting-it) from @vito on slackoverflow could explain the get after put. 

In this particular case, if there is no expectation to use the `put` branch within the build, there should be no implication. However as mentioned in the answer, there is no verification of the push.

If this gonna happen, then it should be well documented. Similar to #327 ,  my question would be for the resource owner: is there a strong need of this feature?

**Comment by vito** [comment]
> The one thing that makes me a bit concerned is that the version created by the `put` would not be detected by a subsequent `check`, as the `check` is still monitoring the `source` branch. I don't totally know the implications of this - maybe @vito can shed some light on whether this is dangerous.

Yeah, this is a good call - it's an awkward asymmetry which will cause the `get` that runs after the `put` to fail because the commit won't exist on the branch that the `get` step fetches.

One way to work around this could be to include the branch in the version (`branch: foo, ref: abcdef`) and modify `in` to respect it rather than using the branch configured in `source`.

A `git-branch` resource type could also make sense but I think it might be a little awkward because branches are created by pushing commits - so it kind of overlaps with `git` functionality. I think once we have Prototypes it would all be handled by a single `git` prototype that supports dealing with both branches and commits (and whatever else).

**Comment by aoldershaw** [comment]
> it's an awkward asymmetry which will cause the get that runs after the put to fail because the commit won't exist on the branch that the get step fetches

It appears this PR gets around that by requiring `source.disable_single_branch = true` when `params.branch` is specified`, which I think should work but is less than ideal. The idea of including `branch` in the version is clever

**Comment by hy0044** [comment]
@aoldershaw @vito
What I want to do is pushing commits to the existing branch (e.g. releng-1.0.0) which is specified dynamically while executing [the release job](https://github.com/concourse/git-resource/pull/333#issuecomment-697222282). It means that we don't need to create new pipeline per release in order to specify the releng branch.

Does the idea of including branch in the version (`branch: foo, ref: abcdef`) means that we can specify `branch: foo` dynamically?

**Comment by hy0044** [comment]
@aoldershaw @vito
I fixed in 392e978 according to the idea of including branch in the version.
Could you please review again?

**Review by aoldershaw — COMMENTED** [review]
Thanks for the changes! It looks pretty nice - just have one suggestion

File: README.md [hunk 1]
```diff
@@ -311,6 +315,9 @@ pushed regardless of the upstream state.
 * `notes`: *Optional.* If this is set then notes will be added to HEAD to the
   `refs/notes/commits` ref. The value should be a path to a file containing the notes.
 
+* `branch`: *Optional.* The branch to push commits.
+  The value should be a path to a file containing the branch.
```
[the remark this answers is not in the document]
Comment (hy0044) [reply README.md]: Thank you for your suggestion.
I fixed in 020f2a1.

**Comment by hy0044** [comment]
@aoldershaw 
I fixed in 020f2a1 according to your suggestion.
Could you please review again?

**Comment by aoldershaw** [comment]
@hy0044 thanks for your prompt changes! I tested it out, and it seems to work. The only thing I'm uncertain about is whether this could lead to unintended behaviour. If you define a `git` resource with `source.branch: master`, and subsequently `put` with `params.branch: releng-1.0.0`, subsequent `get`s to the resource will give back the branch `releng-1.0.0`. Certainly in many cases this is desirable, but I wonder if it may be surprising at times.

For instance, if you had a workflow that runs whenever there's a new commit on `master`, and the workflow pushes to another branch on the same repo, you'd end up with an infinite loop (since pushing to that other branch would still trigger the workflow, even though it's intended to only run on changes to master).

Granted, this could be solved by using two resources - and with this feature, the `put` resource can push to a dynamic branch, which wasn't possible before (without setting a new pipeline). Maybe we should give a heads up in the README saying something like "the version produced by the put step will be picked up by subsequent get steps even if the branch differs from the branch specified in the source"

**Comment by hy0044** [comment]
@aoldershaw
Thank you for pointing out. I didn't realize that infinite loop issue because I usually used two resources of read-only and write-only.
I updated the README in 504a3aa, so could you please check it?

**Pull request merged**